src/SharpBrick.PoweredUp.BlueGigaBLE/BlueGigaBLEPoweredUpBluetoothAdapater.cs

```diff
@@ -0,0 +1,514 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.IO.Ports;
+using SharpBrick.PoweredUp.Bluetooth;
+using System.Linq;
+using System.Globalization;
+using Microsoft.Extensions.Logging;
+using System.Collections.Concurrent;
+using Bluegiga.BLE.Events.Connection;
+using Bluegiga.BLE.Events.ATTClient;
+using Bluegiga.BLE.Responses.ATTClient;
+using System.Text;
+
+namespace SharpBrick.PoweredUp.BlueGigaBLE
+{
+    public class BlueGigaBLEPoweredUpBluetoothAdapater : IPoweredUpBluetoothAdapter, IDisposable, IBlueGigaLogger
+    {
+        #region Properties
+        /// <summary>
+        /// during Connect and Resolve of devices/services/characteristics this status is needed for differentiate the behaviour of end_procedure-handler
+        /// </summary>
+        public BlueGigaBLEStatus BleStatus { get; set; }
+        /// <summary>
+        /// The BGLib-obejct which does the low-level communication over the BlueGiga-adapter
+        /// </summary>
+        public Bluegiga.BGLib Bglib { get; set; }
+        /// <summary>
+        /// The serial connection over which the BleuGiga-adapter communicates
+        /// </summary>
+        public SerialPort SerialAPI { get; set; }
+        /// <summary>
+        /// The ILogger-instance
+        /// </summary>
+        public ILogger _logger { get; set; }
+        /// <summary>
+        /// The devices this adapter knows about (by Discovery or direct Connect)
+        /// </summary>
+        public ConcurrentDictionary<ulong, BlueGigaBLEPoweredUpBluetoothDevice> myDevices;
+        #endregion
+        #region Private Variables
+        private ConcurrentDictionary<ulong, PoweredUpBluetoothDeviceInfo> devicesinfo;
+        private byte[] legoHubServiceBytes;
+        //semaphore for signal successful connection
+        private SemaphoreSlim connected;
+        //semaphore for signal finished discovery of services of a device
+        private SemaphoreSlim allServicesForDeviceDiscovered = new SemaphoreSlim(0, 1);
+        //semaphore for signal finished discovery of characteristics of a service
+        private SemaphoreSlim allCharacteristicsForServiceDiscovered = new SemaphoreSlim(0, 1);
+        private bool wasLastConnectSuccesfull = false;
+        private byte lastConnectionHandle = 0xff;
+        #endregion
+
+        #region Constructor
+        public BlueGigaBLEPoweredUpBluetoothAdapater(String comPortName ,  bool traceDebug, ILogger<SharpBrick.PoweredUp.BlueGigaBLE.BlueGigaBLEPoweredUpBluetoothAdapater> logger = default)
```

Comment:
This happens often to me because I can't remeber the needed class and just start typing the namespace I remeber and then find it by goindf throught the suggestins :-)

Reply:
C# coding guideline always prefers keywords `string` over classes `String`.

Reply:
did so in my last commit (also for byte[], ushort,...